Source/cairo/WaylandCairoSurface.m

```diff
@@ -0,0 +1,204 @@
+
+    surface = cairo_image_surface_create_for_data(window->data,
+						  CAIRO_FORMAT_ARGB32,
+						  window->width,
+						  window->height,
+						  stride);
+
+    window->buffer =
+	wl_shm_pool_create_buffer(pool, 0,
+				  window->width, window->height, stride,
+				  WL_SHM_FORMAT_ARGB8888);
+    wl_shm_pool_destroy(pool);
+
+    close(fd);
+
+    return surface;
+}
+
+@implementation WaylandCairoSurface
+
+- (id) initWithDevice: (void*)device
+{
+    struct window *window = (struct window *) device;
+    NSDebugLog(@"WaylandCairoSurface: initWithDevice win=%d", window->window_id);
+
+    gsDevice = device;
+
+    _surface = create_shm_buffer(window);
+    if (_surface == NULL) {
+	NSDebugLog(@"can't create cairo surface");
+	return 0;
+    }
+
+    wl_surface_attach(window->surface, window->buffer, 0, 0);
+    window->wcs = self;
+
+    return self;
+}
+
+- (void) dealloc
+{
+    struct window *window = (struct window*) gsDevice;
+    NSDebugLog(@"WaylandCairoSurface: dealloc win=%d", window->window_id);
+    [super dealloc];
+}
+
+- (NSSize) size
+{
+    NSDebugLog(@"WaylandCairoSurface: size");
+    struct window *window = (struct window*) gsDevice;
+    return NSMakeSize(window->width, window->height);
+}
+
+- (void) setSurface: (cairo_surface_t*)surface
+{
+    NSDebugLog(@"WaylandCairoSurface: setSurface");
+    _surface = surface;
+}
+
+- (void) handleExposeRect: (NSRect)rect
+{
+    NSDebugLog(@"handleExposeRect");
+    struct window *window = (struct window*) gsDevice;
+    cairo_surface_t *cairo_surface = _surface;
+    double  backupOffsetX = 0;
+    double  backupOffsetY = 0;
+    int x = NSMinX(rect);
+    int y = NSMinY(rect);
+    int width = NSWidth(rect);
+    int height = NSHeight(rect);
+
+    NSDebugLog(@"updating region: %dx%d %dx%d", x, y, width, height);
+
+    if (cairo_surface_status(cairo_surface) != CAIRO_STATUS_SUCCESS)
+    {
+	NSWarnMLog(@"cairo initial window error status: %s\n",
+		   cairo_status_to_string(cairo_surface_status(_surface)));
+    }
+
+    cairo_surface_get_device_offset(cairo_surface, &backupOffsetX, &backupOffsetY);
+    cairo_surface_set_device_offset(cairo_surface, 0, 0);
+
+    cairo_t *cr = cairo_create(cairo_surface);
+    if (width != window->width && 0) {
+	NSDebugLog(@"fake drawing");
+	cairo_set_operator(cr, CAIRO_OPERATOR_SOURCE);
+
+	cairo_rectangle(cr, 0, 0, width, height);
+	cairo_set_source_rgba(cr, 0, 0, 0, 0.8);
+	cairo_fill(cr);
+
+	cairo_rectangle(cr, 10, 10, width - 20, height - 20);
+	cairo_set_source_rgba(cr, 1.0, 0, 0, 1);
+	cairo_fill(cr);
+
+	cairo_select_font_face(cr, "sans",
+			       CAIRO_FONT_SLANT_NORMAL,
+			       CAIRO_FONT_WEIGHT_NORMAL);
+	cairo_set_font_size(cr, 12);
+	cairo_set_source_rgba(cr, 1.0, 1.0, 1.0, 1.0);
+	cairo_move_to(cr, 30, 30);
+	cairo_show_text(cr, "Hello, world!");
+    } else {
+	NSDebugLog(@"real drawing");
+
+	cairo_rectangle(cr, x, y, width, height);
+	cairo_clip(cr);
+	cairo_set_source_surface(cr, cairo_surface, 0, 0);
```

Comment:
Doesn't this copy from cairo_surface to cairo_surface?

Reply:
It does seem to be doing so. I cannot immediately decipher what's the buffer it's supposed to copy from, so in the interest of merging this code, I'm leaving a FIXME.